dvc/parsing/__init__.py

```diff
@@ -453,7 +458,99 @@ def _each_iter(self, key: str) -> "DictStrAny":
             except ContextError as exc:
                 format_and_raise(exc, f"stage '{generated}'", self.relpath)
 
-            # let mypy know that this state is unreachable as format_and_raise
-            # does not return at all (it's not able to understand it for some
-            # reason)
-            raise AssertionError("unreachable")
+
+class MatrixDefinition:
+    def __init__(
+        self,
+        resolver: DataResolver,
+        context: Context,
+        name: str,
+        definition: "DictStrAny",
+        where: str = STAGES_KWD,
+    ):
+        self.resolver = resolver
+        self.relpath = self.resolver.relpath
+        self.context = context
+        self.name = name
+
+        assert DO_KWD in definition
```

Comment:
Is this a leftover or is this meant to serve a purpose?

Reply:
It's a pre-condition, that needs to be fulfilled by the caller.